File: leetcode/649.dota-2-senate.py
#
# @lc app=leetcode id=649 lang=python3
#
# [649] Dota2 Senate
#

# @lc code=start
from queue import Queue

class Solution:
	def predictPartyVictory(self, senate: str) -> str:
		size = len(senate)

		r_arr = [i for i in range(len(senate)) if senate[i] == 'R']
		d_arr = [j for j in range(len(senate)) if senate[j] == 'D']
		while r_arr and d_arr:
			r = r_arr.pop(0)
			d = d_arr.pop(0)
			if r < d:
				r_arr.append(size)
			else:
				d_arr.append(size)
			size += 1
		return "Radiant" if r_arr else "Dire"

# @lc code=end

File: leetcode/649.dota-2-senate.cpp
/*
 * @lc app=leetcode id=649 lang=cpp
 *
 * [649] Dota2 Senate
 */

// @lc code=start
/* class Solution {
public:
	string predictPartyVictory(string senate) {
		queue<char>	q; // maybe queue is for ban?
		int			countR = 0, countD = 0;
		int			i = 0;

		while (true) {
			if (i == senate.size()) {
				if (countR == 0 || countD == 0)
					return (countR ? "Radiant" : "Dire");
				countR = 0;
				countD = 0;
				i = 0;
			}
			if (q.empty() || q.front() == senate[i]) {
				q.push(senate[i]);
				countR += senate[i] == 'R';
				countD += senate[i] == 'D';
				i++;
			}
			else {
				senate.erase(i, 1);
				q.pop();
			}
		}
		return "X";
	}
}; */

class Solution {
public:
	string predictPartyVictory(string senate) {
		queue<int>	d, r;
		int			len = senate.size();

		for (int i = 0; i < len; i++)
			senate[i] == 'R' ? r.push(i) : d.push(i);
		while (!r.empty() && !d.empty()) {
			r.front() > d.front() ? d.push(len++) : r.push(len++);
			r.pop();
			d.pop();
		}
		return (r.empty() ? "Dire" : "Radiant");
	}
};

// @lc code=end
